discovery-provider/src/tasks/user_replica_set.py

```diff
@@ -0,0 +1,285 @@
+                    user_replica_set_events_lookup[user_id]["user"] = user_record
+                    user_replica_set_events_lookup[user_id]["events"].append(event_type)
+                # Process L2 Content Node operations
+                elif event_type == user_replica_set_manager_event_types_lookup['add_or_update_content_node']:
+                    cnode_record = parse_ursm_cnode_record(
+                        update_task,
+                        entry,
+                        cnode_events_lookup[cnode_sp_id]["content_node"]
+                    )
+                    if cnode_record is not None:
+                        cnode_events_lookup[cnode_sp_id]["content_node"] = cnode_record
+                        cnode_events_lookup[cnode_sp_id]["events"].append(event_type)
+            num_user_replica_set_changes += len(user_events_tx)
+
+    # for each record in user_replica_set_events_lookup, invalidate the old record and add the new record
+    # we do this after all processing has completed so the user record is atomic by block, not tx
+    for user_id, value_obj in user_replica_set_events_lookup.items():
+        logger.info(f"user_replica_set.py | Replica Set Processing Adding {value_obj['user']}")
+        invalidate_old_user(session, user_id)
+        session.add(value_obj["user"])
+
+    for content_node_id, value_obj in cnode_events_lookup.items():
+        logger.info(f"user_replica_set.py | Content Node Processing Adding {value_obj['content_node']}")
+        invalidate_old_cnode_record(session, content_node_id)
+        session.add(value_obj["content_node"])
+
+    return num_user_replica_set_changes, user_ids
+
+# Reconstruct endpoint string from primary and secondary IDs
+# Attempt to retrieve from cached values populated in index_network_peers.py
+# If unavailable, then a fallback to ethereum mainnet contracts will occur
+# Note that in the case of an invalid spID - one that is not yet registered on
+# the ethereum mainnet contracts, there will be an empty value in the returned
+# creator_node_endpoint
+# If this discrepancy occurs, a client replica set health check sweep will
+# result in a client-initiated failover operation to a valid set of replicas
+def get_endpoint_string_from_sp_ids(
+        update_task,
+        primary,
+        secondaries,
+        redis
+):
+    sp_factory_inst = None
+    endpoint_string = None
+    primary_endpoint = None
+    try:
+        sp_factory_inst, primary_endpoint = get_endpoint_from_id(
+            update_task,
+            sp_factory_inst,
+            primary
+        )
+        endpoint_string = f"{primary_endpoint}"
+        for secondary_id in secondaries:
+            secondary_endpoint = None
+            sp_factory_inst, secondary_endpoint = get_endpoint_from_id(
+                update_task,
+                sp_factory_inst,
+                secondary_id
+            )
+            if secondary_endpoint:
```

Comment:
noting here per afk discussion

Reply:
test case - set middle replica to an invalid ID and confirm that we get `endpoint1,,endpoint2`

Reply:
without fix, with an invalid spID of 4, note the creator_node_endpoint does _not_ reflect the endpoints. instead of `http://cn3_creator-node_1:4002,,http://cn1_creator-node_1:4000` we have `http://cn3_creator-node_1:4002,http://cn1_creator-node_1:4000`

```
      "creator_node_endpoint": "http://cn3_creator-node_1:4002,http://cn1_creator-node_1:4000",
      "current_user_followee_follow_count": 0,
      "does_current_user_follow": false,
      "followee_count": 0,
      "follower_count": 0,
      "handle": "handle_ideKhR",
      "handle_lc": "handle_idekhr",
      "is_creator": true,
      "is_current": true,
      "is_verified": false,
      "location": "loc_IOGsnR",
      "metadata_multihash": "Qmbt6feamwdFFvPfGHRXnjH8QVg1kVXTCu2XtAGXE7TQp4",
      "name": "name_rQGOkh",
      "playlist_count": 0,
      "primary_id": 3,
      "profile_picture": null,
      "profile_picture_sizes": "QmWud6kcFSUDsbyytGWegjBKS1CJU3Utp35rz8s9D1NWC9",
      "replica_set_update_signer": "0xaD6e9852D2623828A33C9a37FC9ABC4252486018",
      "repost_count": 0,
      "secondary_ids": [
        4,
        1
      ],
```

Reply:
After update, note the same indexing operation has the correctly formatted creator_node_endpoint

```
    {
      "album_count": 0,
      "balance": "0",
      "bio": "i am so cool YmHAnB",
      "blockhash": "0x5d2dfa60a695e0f69919d9bc8030408ed983fc8aab4f506b582c5c2cda95c8d8",
      "blocknumber": 77,
      "cover_photo": null,
      "cover_photo_sizes": "QmTygvnm5QjCxjr53Bgh6aSwvZUQumFFkF6m9NJD6ZV39i",
      "created_at": "2021-02-23T23:44:28 Z",
      "creator_node_endpoint": "http://cn3_creator-node_1:4002,,http://cn1_creator-node_1:4000",
      "current_user_followee_follow_count": 0,
      "does_current_user_follow": false,
      "followee_count": 0,
      "follower_count": 0,
      "handle": "handle_ideKhR",
      "handle_lc": "handle_idekhr",
```

Reply:
great catch on this

Reply:
fix looks great, thanks for documenting